Added a Board representation of the ChArUco board using Pydantic

Rampex1 opened the pull request:
Created a BoardMarker to indicate the ID, and the coordinates of the four corners of the marker.
Created an ArucoBoardSpecification to indicate the list of markers it contains.
Pydantic used for data validation and json serialization.

vaughantnrc commented on src/common/structures/aruco_board_specification.py:
```diff
@@ -0,0 +1,48 @@
+    marker_id: int
+    points: List[Vec3]
+
+    @validator('points', pre=True, each_item=True)
+    def validate_points(cls, value):
+        if isinstance(value, dict):
+            return Vec3(**value)
+        if isinstance(value, Vec3):
+            return value
+        raise ValueError("Invalid type for point")
```
I think you should be ok without this explicit check. pydantic normally won't let you create a data structure with a member whose data type doesn't match what you've specified.

It might make sense to check that `points` is a list of length 4 (assuming we are using ArUco)? If so, then can you also add a tiny test to make sure this validation works as intended?

vaughantnrc commented on src/common/structures/aruco_board_specification.py:
```diff
@@ -0,0 +1,48 @@
+
+class BoardMarker(BaseModel):
+    marker_id: int
+    points: List[Vec3]
```
In the version of Python we're using, we can use lower case `list`.

vaughantnrc commented on src/common/structures/aruco_board_specification.py:
```diff
@@ -0,0 +1,48 @@
+            return value
+        raise ValueError("Invalid type for point")
+
+    def as_dict(self) -> Dict[str, Any]:
+        return {
+            "marker_id": self.marker_id,
+            "points": [point.as_dict() for point in self.points]
+        }
```
Shouldn't need this, pydantic offers a `instance.dict()` function

vaughantnrc commented on src/common/structures/aruco_board_specification.py:
```diff
@@ -0,0 +1,48 @@
+        }
+
+
+class ArucoBoardSpecification(BaseModel):
```
Can we call it a `Board`? Do you see any reason to keep the other words?

vaughantnrc commented on src/common/structures/aruco_board_specification.py:
```diff
@@ -0,0 +1,48 @@
+class ArucoBoardSpecification(BaseModel):
+    board_markers: List[BoardMarker]
+
+    @validator('board_markers', pre=True, each_item=True)
+    def validate_board_markers(cls, value):
+        if isinstance(value, dict):
+            return BoardMarker(**value)
+        if isinstance(value, BoardMarker):
+            return value
+        raise ValueError("Invalid type for board marker")
```
Likewise as comment, above, not sure you need this

vaughantnrc commented on src/common/structures/aruco_board_specification.py:
```diff
@@ -0,0 +1,48 @@
+            return value
+        raise ValueError("Invalid type for board marker")
+
+    def as_dict(self) -> Dict[str, Any]:
+        return {"board_markers": [board_marker.as_dict() for board_marker in self.board_markers]}
```
likewise as above, probably you don't need this.

vaughantnrc reviewed (changes requested):
Thanks, see comments

Rampex1 commented:
Added validator to make sure that the length of the list of points is exactly four.
Removed data type validator since Pydantic does it.
Removed as_dict since Pydantic has its own function as_dict().
Changed class name from ArucoBoardSpecification to Board.

Rampex1 commented:
Also, I should probably have named my commit with MNT.

vaughantnrc commented:
Thank you David
